tests/test_glm.py

```diff
@@ -0,0 +1,1074 @@
+import jax
+import jax.numpy as jnp
+import numpy as np
+import pytest
+import statsmodels.api as sm
+from sklearn.model_selection import GridSearchCV
+
+import nemos as nmo
+
+
+def _test_class_initialization(cls, kwargs, error, match_str):
+    if error:
+        with pytest.raises(error, match=match_str):
+            cls(**kwargs)
+    else:
+        cls(**kwargs)
+
+
+def _test_class_method(cls, method_name, args, kwargs, error, match_str):
+    if error:
+        with pytest.raises(error, match=match_str):
+            getattr(cls, method_name)(*args, **kwargs)
+    else:
+        getattr(cls, method_name)(*args, **kwargs)
+
```

Comment:
I don't love this -- I know it's avoiding repeating code, but there's so many arguments going on here that it makes the calls to this (and thus the tests) much harder to parse. 

Also, there's also this way to have [conditional raises](https://docs.pytest.org/en/7.1.x/example/parametrize.html#parametrizing-conditional-raising):
```python
from contextlib import nullcontext as does_not_raise

if FAIL_CONDITION:
    expectation = pytest.raises(error, match=match_str)
else:
    expectation = does_not_raise()
with expectation:
    code_that_might_fail()
```

This allows you to parametrize it, like in that link. This way you're avoiding re-typing the tested line, which is nice.

For example, your `test_fit_param_length` could be:
```python
  @pytest.mark.parametrize("n_params, expectation", [
      (0, pytest.raises(ValueError, match="Params needs to be array-like of length two.")),
      (1, pytest.raises(ValueError, "Params needs to be array-like of length two.")),
      (2, doest_not_raise()),
      (3, pytest.raises(ValueError, "Params needs to be array-like of length two.")),
  ])
  def test_fit_param_length(self, n_params, expectation, poissonGLM_model_instantiation):
      """
      Test the `fit` method with different numbers of initial parameters.
      Check for correct number of parameters.
      """
      X, y, model, true_params, firing_rate = poissonGLM_model_instantiation
      if n_params == 0:
          init_params = tuple()
      elif n_params == 1:
          init_params = (true_params[0],)
      else:
          init_params = true_params + (true_params[0],) * (n_params - 2)
      with expectation:
          model.fit(X, y, init_params=init_params)
```

Which I think is much clearer, and actually a bit shorter.